File: 数组/归并排序/88. 合并两个有序数组.py

""" 
> 关键是要从后开始遍历
* 从后开始遍历的好处
* 先比较较大的数，把大的数放到数组nums1的后面
* 如果先从较小的数比较，则nums1的元素可能发生移位后挪，时间复杂度高一些。
"""


class Solution(object):
    def merge(self, nums1, m, nums2, n):
        insertIdx = len(nums1) - 1
        nums1Idx = m - 1
        nums2Idx = n - 1
        while insertIdx >= 0:
            if nums1Idx >= 0 and nums2Idx >= 0:
                num1 = nums1[nums1Idx]
                num2 = nums2[nums2Idx]
                if num1 > num2:
                    nums1[insertIdx] = num1
                    nums1Idx -= 1
                else:
                    nums1[insertIdx] = num2
                    nums2Idx -= 1
            elif nums1Idx >= 0:
                # * 此时 nums2 已经遍历完成，只需要处理剩下的 nums1 的元素
                nums1[insertIdx] = nums1[nums1Idx]
                nums1Idx -= 1
            elif nums2Idx >= 0:
                # * 此时 nums1 已经遍历完成，只需要处理剩下的 nums2 的元素
                nums1[insertIdx] = nums2[nums2Idx]
                nums2Idx -= 1

            insertIdx -= 1
        return nums1


Solution().merge(nums1=[1], m=1, nums2=[], n=0)

File: 数组/归并排序/88. 合并两个有序数组.js
/* 
从前往后遍历导致原有数组元素被破坏掉:
nums1=[10, 20, 33, 0, 0, 0] nums2=[1, 2, 3], 从前往后遍历, nums1 的三个值会被 nums2 覆盖掉 

> 关键是要从后开始遍历
* 从后开始遍历的好处
* 先比较较大的数, 把大的数放到数组nums1的后面
* 如果先从较小的数比较, 则nums1的元素可能发生移位后挪, 时间复杂度高一些。

? https://leetcode.cn/problems/merge-sorted-array/solutions/701001/ni-xiang-shuang-zhi-zhen-he-bing-liang-g-ucgj/

 */

var merge = function (nums1, m, nums2, n) {
    let index1 = m - 1;
    let index2 = n - 1;
    let insert = m + n - 1;
    while (index1 >= 0 || index2 >= 0) {
        if (index1 < 0) {
            nums1[insert] = nums2[index2];
            index2--;
        } else if (index2 < 0) {
            nums1[insert] = nums1[index1];
            index1--;
        } else {
            // 这里能够确定 nums1[index1] 和 [index2]都是有确切的值的
            if (nums1[index1] >= nums2[index2]) {
                nums1[insert] = nums1[index1];
                index1--;
            } else {
                nums1[insert] = nums2[index2];
                index2--;
            }
        }

        insert--;
    }
};
